### PR TITLE
fix: `improve _extract_tables_from_sql`

### PR DESCRIPTION
**Pull request by betodealmeida**
### SUMMARY

Improve `_extract_tables_from_sql` from https://github.com/apache/superset/pull/26476 so that a proper exception is returned if the SQL fails to parse when extracting tables mentioned in a query. The previous behavior was to return an empty set of tables in the query.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF

N/A

### TESTING INSTRUCTIONS

Updated unit tests.

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset/sql_parse.py [hunk 1]
```diff
@@ -287,9 +292,15 @@ def _extract_tables_from_sql(self) -> set[Table]:
...
-        except ParseError:
-            logger.warning("Unable to parse SQL (%s): %s", self._dialect, self.sql)
-            return set()
+        except ParseError as ex:
+            dialect = self._dialect or "generic"
+            raise SupersetSecurityException(
+                SupersetError(
+                    error_type=SupersetErrorType.QUERY_SECURITY_ACCESS_ERROR,
+                    message=__(f"Unable to parse SQL ({dialect}): {self.sql}"),
+                    level=ErrorLevel.ERROR,
```
Comment (eschutho) [inline superset/sql_parse.py:301]: @betodealmeida is this a user error or system error?

Comment (betodealmeida) [reply superset/sql_parse.py:301]: It's hard to tell. It could be invalid SQL (user error), or it could be because limitations of the parser (system error).

Comment (eschutho) [reply superset/sql_parse.py:301]: Ok, that makes sense. I'm thinking of a scenario where someone is making a lot invalid sql calls and it just keeps loading up a system monitoring queue with what looks to be an outage of a sort.

Comment (betodealmeida) [reply superset/sql_parse.py:301]: Yeah, I'm worried about that too, and I think that's why originally the behavior was to not to fail. But the alternative is allowing people to run queries on tables they don't have access because we fail to extract the reference from the SQL.

Comment (john-bodley) [reply superset/sql_parse.py:302]: @betodealmeida I _think_ having Superset being less lenient/blasé is likely a good thing, i.e., previously it blindly would could have* given a user access whenever it wasn't able to parse the SQL, irrespective of whether the user had access to the underly data.

The major challenge I see is if SQLGlot wrongfully isn't able to parse a SQL statement and thus we're wrongfully blocking a query. How do we help unblock users here?

\* This is dependent on how one configured their security manager.

**Comment by john-bodley** [comment]
The TL;DR is  (per learnings during my internal postmortem writeup) `sqlparse` is a non-validating SQL parser whereas SQLGlot is a pseudo-validating* SQL parser. I'm not saying one is right or wrong, but sqlparse is definitely more lenient which could be perceived as beneficial for out specific use case, i.e., trying to identify the tables that a SQL query (be it valid or not) is referencing.

\* SQLGlot doesn't claim to validate SQL (per its [README.md](https://github.com/tobymao/sqlglot/blob/main/README.md)), 

> SQLGlot does not aim to be a SQL validator - it is designed to be very forgiving

it does seem to be somewhat opinionated (understandably so, i.e., if you're unable to successfully parse token(s) anything downstream in the AST is likely deemed questionable).

**Review by john-bodley — APPROVED** [review]
